modules/infra/api/nexthop.c

```diff
@@ -39,50 +39,77 @@ static struct gr_api_handler config_set_handler = {
 	.callback = nh_config_set,
 };
 
-static struct api_out nh_add(const void *request, void ** /*response*/) {
-	const struct gr_nh_add_req *req = request;
-	const struct nexthop_af_ops *ops;
-	struct gr_nexthop base = req->nh;
-	struct nexthop *nh = NULL;
+static int nh_add_blackhole(struct gr_nexthop *base) {
+	base->state = GR_NH_S_REACHABLE;
+	base->flags |= GR_NH_F_STATIC;
+	base->af = GR_AF_UNSPEC;
+	base->iface_id = GR_IFACE_ID_UNDEF;
+	return 0;
+}
+
+static int nh_add_l3(struct gr_nexthop *base) {
 	struct iface *iface;
-	int ret;
 
-	base.flags = 0;
-	switch (base.af) {
+	switch (base->af) {
 	case GR_AF_IP4:
-		if (base.ipv4 == 0)
-			return api_out(EDESTADDRREQ, 0);
-
+		if (base->ipv4 == 0)
+			return -EDESTADDRREQ;
 		break;
 	case GR_AF_IP6:
-		if (rte_ipv6_addr_is_unspec(&base.ipv6))
-			return api_out(EDESTADDRREQ, 0);
+		if (rte_ipv6_addr_is_unspec(&base->ipv6))
+			return -EDESTADDRREQ;
 
 		break;
 	case GR_AF_UNSPEC:
-		if (base.ipv4 || !rte_ipv6_addr_is_unspec(&base.ipv6))
-			return api_out(EINVAL, 0);
+		if (base->ipv4 || !rte_ipv6_addr_is_unspec(&base->ipv6))
+			return -EINVAL;
 
-		base.flags |= GR_NH_F_LINK | GR_NH_F_STATIC;
+		base->flags |= GR_NH_F_LINK | GR_NH_F_STATIC;
 		break;
 	default:
-		return api_out(ENOPROTOOPT, 0);
+		return -ENOPROTOOPT;
 	}
 
-	iface = iface_from_id(base.iface_id);
+	iface = iface_from_id(base->iface_id);
 	if (iface == NULL)
-		return api_out(errno, 0);
+		return -errno;
+
+	base->vrf_id = iface->vrf_id;
+	base->state = GR_NH_S_NEW;
+	if (!rte_is_zero_ether_addr(&base->mac)) {
+		if (base->af == GR_AF_UNSPEC)
+			return -EINVAL;
+
+		base->state = GR_NH_S_REACHABLE;
+		base->flags |= GR_NH_F_STATIC;
+	}
+	return 0;
+}
 
-	base.vrf_id = iface->vrf_id;
-	base.type = GR_NH_T_L3;
-	base.state = GR_NH_S_NEW;
-	if (!rte_is_zero_ether_addr(&base.mac)) {
-		if (base.af == GR_AF_UNSPEC)
-			return api_out(EINVAL, 0);
+static struct api_out nh_add(const void *request, void ** /*response*/) {
+	const struct gr_nh_add_req *req = request;
+	const struct nexthop_af_ops *ops;
+	struct gr_nexthop base = req->nh;
+	struct nexthop *nh = NULL;
+	int ret;
 
-		base.state = GR_NH_S_REACHABLE;
-		base.flags |= GR_NH_F_STATIC;
+	base.flags = 0;
+	switch (base.type) {
+	case GR_NH_T_BLACKHOLE:
+	case GR_NH_T_REJECT:
+		ret = nh_add_blackhole(&base);
+		break;
+	case GR_NH_T_L3:
+	case GR_NH_T_SR6_OUTPUT:
+	case GR_NH_T_SR6_LOCAL:
+	case GR_NH_T_DNAT:
+		ret = nh_add_l3(&base);
+		break;
+	default:
+		return api_out(EINVAL, 0);
 	}
+	if (ret)
+		return api_out(ret, 0);
```

Comment:
nh_add_blackhole and nh_add_l3 return negative values. You should check accordingly:

```c
if (ret < 0)
         return api_out(-ret, 0);
```